mr/src/main/java/org/opensearch/hadoop/rest/commonshttp/CommonsHttpTransport.java

```diff
@@ -712,12 +793,95 @@ public Object run() throws Exception {
             headerValues.add(responseHeader.getValue());
         }
 
-        // the request URI is not set (since it is retried across hosts), so use the http info instead for source
+        // the request URI is not set (since it is retried across hosts), so use the
+        // http info instead for source
         return new SimpleResponse(http.getStatusCode(), new ResponseInputStream(http), httpInfo, headers);
     }
 
+    private String queryParamsString(Multimap<String, String> queryParams) {
+        final ImmutableList.Builder<String> result = ImmutableList.builder();
+        for (Map.Entry<String, Collection<String>> param : new TreeMap<>(queryParams.asMap()).entrySet()) {
+            for (String value : param.getValue()) {
+                result.add(SdkHttpUtils.urlEncode(param.getKey()) + '=' + SdkHttpUtils.urlEncode(value));
+            }
+        }
+
+        return '?' + AMPERSAND_JOINER.join(result.build());
+    }
+
+    private void awsSigV4SignRequest(Request request, HttpMethod http, byte[] bodyBytes)
```

Comment:
could we reuse some work in https://github.com/opensearch-project/opensearch-java/blob/main/java-client/src/main/java/org/opensearch/client/transport/aws/AwsSdk2Transport.java.